reactos/drivers/usb/usbohci_new/usbohci.h

```diff
@@ -0,0 +1,253 @@
+#ifndef USBOHCI_H__
+#define USBOHCI_H__
+
+#include <ntddk.h>
+#include <windef.h>
+#include <stdio.h>
+#include <hubbusif.h>
+#include <usbbusif.h>
+#include <usbdlib.h>
+#include <drivers/usbport/usbmport.h>
+
+#include "hardware.h"
+
+extern USBPORT_REGISTRATION_PACKET RegPacket;
+
+#define OHCI_HCD_ED_FLAG_CONTROLL      0x00000001
+#define OHCI_HCD_ED_FLAG_RESET_ON_HALT 0x00000008
+#define OHCI_HCD_ED_FLAG_NOT_ACCESSED  0x00000010
+
+#define OHCI_HCD_TD_FLAG_ALLOCATED       0x00000001
+#define OHCI_HCD_TD_FLAG_PROCESSED       0x00000002
+#define OHCI_HCD_TD_FLAG_CONTROLL_STATUS 0x00000004
+#define OHCI_HCD_TD_FLAG_DONE            0x00000008
+#define OHCI_HCD_TD_FLAG_NOT_ACCESSED    0x00000010
+
+#define OHCI_MAX_ISO_TRANSFER_SIZE        0x10000
+#define OHCI_MAX_CONTROL_TRANSFER_SIZE    0x10000
+#define OHCI_MAX_BULK_TRANSFER_SIZE       0x40000
+#define OHCI_MAX_INTERRUPT_TRANSFER_SIZE  0x1000
+
+#define OHCI_MAX_ISO_TD_COUNT        64
+#define OHCI_MAX_CONTROL_TD_COUNT    38
+#define OHCI_MAX_BULK_TD_COUNT       68
+#define OHCI_MAX_INTERRUPT_TD_COUNT  4
+
+#define OHCI_TRANSFER_FLAGS_SHORT_TRANSFER_OK  1
+
+typedef struct _OHCI_TRANSFER *POHCI_TRANSFER;
+
+typedef union _OHCI_HW_TRANSFER_DESCRIPTOR {
+  struct {
+    OHCI_TRANSFER_DESCRIPTOR gTD; // must be aligned to a 16-byte boundary
+    USB_DEFAULT_PIPE_SETUP_PACKET SetupPacket;
+    ULONG Padded[2];
+  };
+  struct {
+    OHCI_ISO_TRANSFER_DESCRIPTOR iTD; // must be aligned to a 32-byte boundary
+  };
+} OHCI_HW_TRANSFER_DESCRIPTOR, *POHCI_HW_TRANSFER_DESCRIPTOR;
+
+C_ASSERT(sizeof(OHCI_HW_TRANSFER_DESCRIPTOR) == 32);
+
+typedef struct _OHCI_HCD_TD {
+  // Hardware part
+  OHCI_HW_TRANSFER_DESCRIPTOR HwTD; // must be aligned to a 32-byte boundary
+  // Software part
+  ULONG_PTR PhysicalAddress;
+  ULONG Flags;
+  POHCI_TRANSFER OhciTransfer;
+  struct _OHCI_HCD_TD * NextHcdTD;
+  ULONG TransferLen;
+  LIST_ENTRY DoneLink;
+  ULONG Pad[1];
+} OHCI_HCD_TD, *POHCI_HCD_TD;
+
+C_ASSERT(sizeof(OHCI_HCD_TD) == 0x40);
+
+typedef struct _OHCI_HCD_ED {
+  // Hardware part
+  OHCI_ENDPOINT_DESCRIPTOR HwED; // must be aligned to a 16-byte boundary
+  // Software part
+  ULONG_PTR PhysicalAddress;
+  ULONG Flags;
+  LIST_ENTRY HcdEDLink;
+  ULONG Pad[8];
+} OHCI_HCD_ED, *POHCI_HCD_ED;
+
+C_ASSERT(sizeof(OHCI_HCD_ED) == 0x40);
+
+#define OHCI_STATIC_ED_TYPE_INTERRUPT  0
+#define OHCI_STATIC_ED_TYPE_CONTROLL   1
+#define OHCI_STATIC_ED_TYPE_BULK       2
+
+typedef struct _OHCI_STATIC_ED {
+  // Software only part
+  POHCI_ENDPOINT_DESCRIPTOR HwED;
+  ULONG_PTR PhysicalAddress;
+  UCHAR HeadIndex;
+  UCHAR Reserved[3];
+  LIST_ENTRY Link;
+  ULONG Type;
+  PULONG pNextED;
+  ULONG HccaIndex;
+} OHCI_STATIC_ED, *POHCI_STATIC_ED;
+
+typedef struct _OHCI_HC_RESOURCES {
+  OHCI_HCCA HcHCCA; // (256 byte align)
+  OHCI_ENDPOINT_DESCRIPTOR InterrruptHeadED[63]; // (16 byte align)
```

Comment:
If alignment is important, try C_ASSERT on offsetof%16 == 0 (not sure if that is possible in a C_ASSERT)

Reply:
Host Controller Resouses begin at the beginning of the memory page. And if sizeof (HcHCCA) == 256, then the alignment will be respected.
A sizeof (HcHCCA) will be 256, because there is C_ASSERT (sizeof (OHCI_HCCA) == 256).